memcached.c

```diff
@@ -4181,6 +4263,740 @@ static inline void process_get_command(conn *c, token_t *tokens, size_t ntokens,
+        pthread_mutex_lock(&c->thread->stats.mutex);
+        if (ttl_set) {
+            c->thread->stats.touch_cmds++;
+            c->thread->stats.touch_misses++;
+        } else {
+            c->thread->stats.get_misses++;
+            c->thread->stats.get_cmds++;
+        }
+        MEMCACHED_COMMAND_GET(c->sfd, key, nkey, -1, 0);
+        pthread_mutex_unlock(&c->thread->stats.mutex);
+
+        out_string(c, "END");
+    }
+    return;
+error:
+    if (it && of.locked) {
+        item_unlock(hv);
+    }
+    out_errstring(c, errcode, errstr, key, nkey);
+}
+
+static void process_mset_command(conn *c, token_t *tokens, const size_t ntokens) {
+    char *key;
+    size_t nkey;
+    unsigned int flags = 0;
+    uint32_t exptime_int = 0;
+    uint64_t req_cas_id = 0;
+    time_t exptime;
+    int vlen = 0;
+    int rtokens;
+    item *it;
+    char *opts;
+    int olen, i;
+    short comm = NREAD_SET;
+    struct _meta_flags of = {0}; // option bitflags.
+    char *errcode = "CLIENT_ERROR";
+    char *errstr = "bad command line format";
+    uint32_t hv;
+
+    assert(c != NULL);
+
+    // TODO: most of this is identical to mget.
+    if (tokens[KEY_TOKEN].length > KEY_MAX_LENGTH) {
+        out_string(c, "CLIENT_ERROR bad command line format");
+        return;
+    }
+
+    key = tokens[KEY_TOKEN].value;
+    nkey = tokens[KEY_TOKEN].length;
+
+    rtokens = ntokens - 3; // cmd, key, final.
+
+    if (rtokens == 0) {
+        out_string(c, "CLIENT_ERROR bad command line format");
+        return;
+    }
+
+    opts = tokens[KEY_TOKEN + 1].value;
+    olen = tokens[KEY_TOKEN + 1].length;
+    rtokens--;
+
+    if (olen > MFLAG_MAX_OPT_LENGTH) {
+        out_string(c, "CLIENT_ERROR options flags too long");
+        return;
+    }
+
+    // scrubs duplicated options and sets flags for how to load the item.
+    // TODO: I, E, APL?
+    rtokens -= _meta_flag_preparse(opts, olen, &of);
+
+    // FIXME: make string more clear and add key to response
+    if (rtokens < 0) {
+        out_string(c, "CLIENT_ERROR not enough tokens supplied");
+        return;
+    }
+
+    // Set noreply after tokens are generally understood to be clear to users.
+    c->noreply = of.no_reply;
+    rtokens = KEY_TOKEN + 2;
+
+    // TODO: Do we need to return the tokens?
+    // since NOT_STORED and so on are far away from this (past a read), if we
+    // need to store the tokens it would have to be stowed with the connection
+    // structure
+
+    for (i = 0; i < olen; i++) {
+        switch (opts[i]) {
+            case 'F':
+                if (!safe_strtoul(tokens[rtokens].value, &flags)) {
+                    goto error;
+                }
+                rtokens++;
+                break;
+            case 'C':
+                if (!safe_strtoull(tokens[rtokens].value, &req_cas_id)) {
+                    goto error;
+                }
+                comm = NREAD_CAS;
+                rtokens++;
+                break;
+            case 'S':
```

Comment:
Is it useful to have the mset without `S` ? Looks like we are truncating (`vlen=0`) the item in that case.

Reply:
People set zero byte items sometimes. It could be a shortcut or some other thing; typically not useful no.